kernel/kernel-api/src/main/java/io/delta/kernel/internal/stats/FileSizeHistogram.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (2024) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.stats;
+
+import static io.delta.kernel.internal.util.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.util.*;
+
+/** A histogram that tracks file size distributions and their counts. */
+public class FileSizeHistogram {
+  private static final long KB = 1024;
+  private static final long MB = KB * 1024;
+  private static final long GB = MB * 1024;
+
+  private static final List<Long> DEFAULT_BIN_BOUNDARIES = createDefaultBinBoundaries();
+
+  // TODO: Add final after ToRow() implemented.
+  final List<Long> sortedBinBoundaries;
+  final Long[] fileCounts;
+  final Long[] totalBytes;
+
+  // A sorted array of bin boundaries where each element represents the start of a bin (inclusive)
+  // and the next element represents the end of the bin (exclusive). The first element must be 0.
+  private static List<Long> createDefaultBinBoundaries() {
+    return Arrays.asList(
+        0L,
+        // Power of 2 till 4 MB
+        8 * KB,
+        16 * KB,
+        32 * KB,
+        64 * KB,
+        128 * KB,
+        256 * KB,
+        512 * KB,
+        1 * MB,
+        2 * MB,
+        4 * MB,
+        // 4 MB jumps till 40 MB
+        8 * MB,
+        12 * MB,
+        16 * MB,
+        20 * MB,
+        24 * MB,
+        28 * MB,
+        32 * MB,
+        36 * MB,
+        40 * MB,
+        // 8 MB jumps till 120 MB
+        48 * MB,
+        56 * MB,
+        64 * MB,
+        72 * MB,
+        80 * MB,
+        88 * MB,
+        96 * MB,
+        104 * MB,
+        112 * MB,
+        120 * MB,
+        // 4 MB jumps till 144 MB
```

Comment:
Let's include the remaining comment from delta-spark `(since we want more detail around the 128 MB mark)`